src/wayland/seat/touch.rs

```diff
@@ -0,0 +1,184 @@
+    pub fn up(&self, time: u32, slot: TouchSlot) {
+        self.inner.borrow_mut().up(time, slot);
+    }
+
+    /// Notify clients about touch motion.
+    pub fn motion(&self, time: u32, slot: TouchSlot, location: Point<f64, Logical>) {
+        self.inner.borrow_mut().motion(time, slot, location);
+    }
+
+    /// Notify clients about touch shape changes.
+    pub fn shape(&self, slot: TouchSlot, major: f64, minor: f64) {
+        self.inner.borrow_mut().shape(slot, major, minor);
+    }
+
+    /// Notify clients about touch shape orientation.
+    pub fn orientation(&self, slot: TouchSlot, orientation: f64) {
+        self.inner.borrow_mut().orientation(slot, orientation);
+    }
+
+    /// Notify clients about touch cancellation.
+    ///
+    /// This should be sent by the compositor when the currently active touch
+    /// slot was recognized as a gesture.
+    pub fn cancel(&self) {
+        self.inner.borrow_mut().cancel();
+    }
+}
+
+/// Touch-slot focused Wayland client state.
+#[derive(Default, Debug)]
+struct TouchFocus {
+    surface_offset: Point<f64, Logical>,
+    handles: Vec<WlTouch>,
+}
+
+#[derive(Default, Debug)]
+struct TouchInternal {
+    known_handles: Vec<WlTouch>,
+    focus: HashMap<TouchSlot, TouchFocus>,
+}
+
+impl TouchInternal {
+    fn down(
+        &mut self,
+        time: u32,
+        surface: &WlSurface,
+        surface_offset: Point<i32, Logical>,
+        slot: TouchSlot,
+        location: Point<f64, Logical>,
+    ) {
+        // Update focused client state.
+        let focus = self.focus.entry(slot).or_default();
+        focus.surface_offset = surface_offset.to_f64();
+        focus.handles.clear();
+
+        // Select all WlTouch instances associated to the active WlSurface.
+        for handle in &self.known_handles {
+            if handle.as_ref().same_client_as(surface.as_ref()) {
+                focus.handles.push(handle.clone());
+            }
+        }
+
+        let (x, y) = (location - focus.surface_offset).into();
+        let serial = SERIAL_COUNTER.next_serial().into();
+        self.with_focused_handles(slot, |handle| {
+            handle.down(serial, time, surface, slot.into(), x, y)
+        });
+    }
+
+    fn up(&self, time: u32, slot: TouchSlot) {
+        let serial = SERIAL_COUNTER.next_serial().into();
+        self.with_focused_handles(slot, |handle| handle.up(serial, time, slot.into()));
+    }
+
+    fn motion(&self, time: u32, slot: TouchSlot, location: Point<f64, Logical>) {
+        let focus = match self.focus.get(&slot) {
+            Some(slot) => slot,
+            None => return,
+        };
+
+        let (x, y) = (location - focus.surface_offset).into();
+        self.with_focused_handles(slot, |handle| handle.motion(time, slot.into(), x, y));
+    }
+
+    fn shape(&self, slot: TouchSlot, major: f64, minor: f64) {
+        self.with_focused_handles(slot, |handle| handle.shape(slot.into(), major, minor));
```

Comment:
This is missing a version check, shape is available since version 6

Reply:
Do you know if there's some minimum level Smithay assumes to be present? Because I noticed release is only part of version 3 but I don't think Smithay ever checks version for that?

Reply:
> Do you know if there's some minimum level Smithay assumes to be present? Because I noticed release is only part of version 3 but I don't think Smithay ever checks version for that?

That would be a bug then. Compositors are required to support all previous versions for stable protocols.

Reply:
I think Smithay gets around this by never actually handling release but just using automatic destructors using `assign_destructor`. So I think the existing implementations are okay? I just noticed that there weren't any version checks for release.